[WMS][12.0] - stock_location_zone - alpha version

jbaudoux opened the pull request:
This is a first prototype of the part "Warehouse map and location attributes" RFC: https://github.com/OCA/stock-logistics-warehouse/issues/654

https://docs.google.com/document/d/1mct6bFFWJqW01wGFcjc-uQNEjyCxvh6Y9TuFdRhe-b0/edit#heading=h.49w4ly4e5y8g

jbaudoux committed:
[ADD] stock_location_zone

jbaudoux committed:
[ADD] stock_location_zone

jbaudoux committed:
Refactoring

guewen commented on stock_location_zone/models/stock_location.py:
```diff
@@ -0,0 +1,78 @@
+    kind = fields.Selection([
+        ('zone', 'Picking Zone'),
+        ('area', 'Area'),
+        ('bin', 'Bin'),
```
Can't we compute this based on other properties? Example, a leaf is a bin, a record with a `picking_zone_id` is a zone, an area I don't know... Maybe not possible but worth to ask :)

guewen replied:
BTW the vertical lift ones are computed, so a manual selection as here won't really fit together

jbaudoux replied:
I would like to be able to use the tray/cell concepts you implemented without having a vertical lift (shuttle). I could have a table or shelf with x/y coordinates that could use the matrix display.

guewen replied:
I see, makes sense

guewen commented on stock_location_zone/models/stock_location.py:
```diff
@@ -0,0 +1,78 @@
+            area = location
+            while not area.location_name_format:
+                if not area.location_id:
+                    continue
```
```suggestion
                    break
```
`continue` will loop infinitely

jbaudoux replied:
return

guewen commented on stock_location_zone/models/stock_location.py:
```diff
@@ -0,0 +1,78 @@
+                if not area.location_id:
+                    continue
+                area = area.location_id
+            location.name = area.location_name_format\
```
We have to check if `area.location_name_format` is not empty, in case no location in the hierarchy has one

jbaudoux replied:
then let's return instead of break

guewen replied:
Nope, return would skip the next records in `self`

guewen commented on stock_location_zone/models/picking_zone.py:
```diff
@@ -0,0 +1,25 @@
+
+
+class PickingZone(models.Model):
+    _name = 'picking.zone'
```
The model is bothering me a bit, I would expect `stock.picking.zone` to be coherent with `stock.location`, `stock.picking.type`, ...

jbaudoux replied:
me too. I agree. Should I then add migration file as it was named like that on v10 projects

guewen commented on setup/stock_location_zone/odoo/addons/stock_location_zone:
```diff
@@ -0,0 +1 @@
+../../../stock_location_zone
```
```suggestion
../../../../stock_location_zone
```

guewen commented on stock_location_zone/models/stock_location.py:
```diff
@@ -0,0 +1,78 @@
+    location_name_format = fields.Char(
+        'Location Name Format',
+        help="Format string that will compute the name of the location. "
+             "Use 'self' to access location object. Example: "
```
This is too unsafe, allowing too much python code, we should allow a safe list of keywords

jbaudoux replied:
Changed object argument to **object.read()

guewen replied:
It fails with the following error on a new record:

```
 The requested operation cannot be completed due to record rules: Document type: stock.location, Operation: read, Records: <odoo.models.NewId object at 0x7f659f6921c8>, User: 2
```

guewen commented on stock_location_zone/models/stock_location.py:
```diff
@@ -0,0 +1,78 @@
+    _sql_constraints = [
+        (
+            'unique_location_name',
+            'UNIQUE(name, location_id)',
```
name must be reset on copy or this constraint prevents duplicating a location

jgrandguillaume commented on stock_location_zone/models/stock_location.py:
```diff
@@ -0,0 +1,78 @@
+    kind = fields.Selection([
+        ('zone', 'Picking Zone'),
+        ('area', 'Area'),
+        ('bin', 'Bin'),
```
```suggestion
        ('bin', 'Bin'), ],
```

jgrandguillaume commented on stock_location_zone/models/stock_location.py:
```diff
@@ -0,0 +1,78 @@
+        ('zone', 'Picking Zone'),
+        ('area', 'Area'),
+        ('bin', 'Bin'),
+        ],
```
```suggestion
```

guewen commented:
I'll push a commit

Commit:
Add corrections to stock_location_zone

* Allow copy of stock locations (was blocked by constraint on unique name)
* Correct loop in _compute_name returning too early if a record had no
parent with a 'location_name_format'
* Rename field pick_type_id to picking_type_id for coherency
* Add missing _description on stock.picking.zone
* Correct location_name_format format when the record is a NewId

Commit:
fixup! Add corrections to stock_location_zone

Commit:
Replace SQL unique constraint by a python one

The new constraints prevents having 2 locations with the same name in
the same zone. The SQL UNIQUE constraint was blocking the installation
of some addons due to demo data creating, by default, some locations
with the same name for the warehouses' locations.

Alternative: we could create an INDEX UNIQUE (name, location_id) WHERE
picking_zone_id IS NOT NULL
But it depends if we want to check the whole tree or not.

guewen commented on stock_location_zone/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+        'views/stock_location.xml',
+        'security/ir.model.access.csv',
+    ],
+    'installable': True,
```
```suggestion
    'installable': True,
    'development_status': 'Alpha',
```

jgrandguillaume reviewed (approved):
LGTM, functional test ok. To add the development status alpha and ready to merge

Commit:
Use unique index for unique location name per zone

The python one was much too slow with large datasets. Here we only
check one level (only when the zone is directly assigned), the previous
SQL constraint was checking the locations of the same level only anyway.

Commit:
Set Alpha development status

* Add copyright header
* Add myself as contributor of stock_location_zone
* Fix lint

grindtildeath reviewed (approved):
Approved to add more changes later.

jgrandguillaume commented:
/ocabot merge

jgrandguillaume commented:
/ocabot merge

I try again as Travis was red because of unaccessibility...

rousseldenis commented:
@jbaudoux Sorry arriving late in this.

Just a comment about sql constraint with WHERE.

Actually, you can use _sql_constraints. See what I did there https://github.com/OCA/partner-contact/blob/10.0/partner_bank_sort_code/models/res_bank.py#L14

You can use sql clause EXCLUDE.

See https://www.postgresql.org/docs/11/ddl-constraints.html#DDL-CONSTRAINTS-EXCLUSION

 or https://www.cybertec-postgresql.com/en/postgresql-exclude-beyond-unique/

grindtildeath commented:
@rousseldenis Thanks for the hint, anyway I took care of this in #711 where it's not needed after simplification 

EDIT: Was still needed and your solution works like a charm :)
